### DISCUSSION TITLE
Snap version is wrong for connection with MetaMask

### DISCUSSION THREAD

**Discussion by Jacopodd** [blockchain-lab-um/ssi-snap#608] [General]
Hi, I'm using this Masca to test a system I created for a university project. After starting the website locally, I try to connect Masca with MetaMask but it gives me the following error message:
Cannot install version 1.2.0 of snap "npm:@blockchain-lab-um/masca": The snap is not on the allowlist. 
Do you know how to solve it?

**Comment by pseudobun** [comment]
Hi @Jacopodd!

TLDR:
- For Masca versions `^1.1.0`, use [MetaMask Flask](https://chromewebstore.google.com/detail/metamask-flask-developmen/ljfoeinjpaedjfecbmggjgodbgkmjkjk).
- For Masca versions up to `1.1.0`, use the regular production MetaMask.

For development purposes, we recommend using [MetaMask Flask (development build)](https://chromewebstore.google.com/detail/metamask-flask-developmen/ljfoeinjpaedjfecbmggjgodbgkmjkjk), which is unrestricted in terms of snaps. It allows developers to install and use any snap regardless of whether it has been approved by MetaMask or not. In the production build of MetaMask, the one you are probably using right now, the snap and each of its new versions must be approved by MetaMask. Snaps and their corresponding versions currently available in MetaMask can be found in [the snaps registry](https://snaps.metamask.io/), specifically [Masca](https://snaps.metamask.io/snap/npm/blockchain-lab-um/masca/). As you can see, our latest stable version available in MetaMask is `1.1.0`. For all our released versions, you can take a look at the [npm registry](https://www.npmjs.com/package/@blockchain-lab-um/masca?activeTab=versions) and can install them to MetaMask Flask.

**Comment by martines3000** [comment]
@Jacopodd I recommend you downgrade the `masca-connector` version to `1.2.1` for now. This way you avoid any incompatibilities between the Masca and Masca-connector versions.

Masca v1.2.2 will soon be on the allowlist, it is currently under review by MetaMask.

If you want to test our latest features and improvements, follow the advice from @pseudobun and install the development version of MetaMask (MetaMask Flask).

I hope this information helps 👍

**Comment by Jacopodd** [comment]
Perfect, Flusk works correctly with MetaMask.
Now I have another problem, I should scan some QrCodes to test what I said before, but the PC doesn't have the ability to scan (and if I load the QrCode it says that it is invalid even though it is generated by the official EBSI website).
When I go to "This Device can't scan/upload QrCode" it doesn't let me create the connection because when I click on the button to authenticate with Ethereum nothing happens. Do you have advice?

**Reply by martines3000** [reply]
Hi. Currently the "Sign In" flow is broken, we are working on a fix. We will probably push the updates by end of tomorrow. I will let you know.

The uploading of QR Codes also doesn't work correctly. Thanks for reporting the issue.

**Reply by martines3000** [reply]
@Jacopodd Hi. I fixed the issue related to the `sign in` on our dapp and redeployed it.

The fixes should be available and working on https://masca.io.

If you encounter any more issues or you have feedback for us, please let us know. We are trying to improve our products as much as possible 🙂.

**Reply by Jacopodd** [reply]
@martines3000 
Hi, by accessing the site: https://masca.io
and clicking on "Launch App", it does not connect and gives the following error: Snap "npm:@blockchain-lab-um/masca@1.2.0" is already installed. Couldn't update to a version inside requested "1.1.0" range.

While if I log in to localhost:3000, login with Eterhum doesn't work but it connects to MetaMask Flusk. I also found a way using the smartphone to scan the QrCode but once scanned, the first, second and third steps are complete, but the fourth step "Start Flow" remains blocked and nothing happens. Can you advise me how to continue?

**Reply by martines3000** [reply]
Snaps can not be downgraded (only updated to higher versions). If you delete the snap here:
![image](https://github.com/blockchain-lab-um/masca/assets/35891136/9ba03337-1685-489b-96fc-09a8b39cfd28)

You will be able to install the v1.1.0 version.

**Reply by martines3000** [reply]
@Jacopodd Try the QR Code scanning on https://masca.io after installing Masca v1.1.0 and if the last part of the flow still doesn't work, please let me know and I will try to help you debug it further.

**Reply by Jacopodd** [reply]
I went to the site https://masca.io
I clicked on "Launch App", established the connection with Etherium and also connected the smartphone as you told me. Everything works. I scan the QrCode on the official EBSI test website and click on "Start Flow". I accept the credential offer on MetaMask Flow but it gives me the error message: failed to finish CredentialOffer flow.
![Screenshot (113)](https://github.com/blockchain-lab-um/masca/assets/80394163/03ec98ae-430f-4694-8653-55d108261cb4)

**Reply by martines3000** [reply]
@Jacopodd Which flow (QR Code) are you testing it with ?

**Reply by martines3000** [reply]
We only support Holder Wallet Tests (should work with all 4 test cases)
![image](https://github.com/blockchain-lab-um/masca/assets/35891136/db3a8800-366d-40a6-b89c-e81fe9e13d07)

**Reply by martines3000** [reply]
Also note you need to use `did:key (EBSI)` as the did method. 
![image](https://github.com/blockchain-lab-um/masca/assets/35891136/167a81ff-4ff1-4af6-86cb-5b226f6139f3)

**Reply by Jacopodd** [reply]
yes, I selected did:key (ebsi) and pasted the did on the ebsi test site.  I tried scanning the QR code generated by In time credential and it gives me the error I wrote to you before

**Reply by martines3000** [reply]
@Jacopodd Hi. I tried the same steps you just described and it worked for me. Could you maybe provide more details how you did it ? Maybe a short GIF or video would help, as it will be easier for me to find the issue.

**Reply by Jacopodd** [reply]
@martines3000 Il video che ho registrato è troppo grande, potete mandarmi una email dove inviare il video?

**Reply by martines3000** [reply]
Sure. Send it to `domajnko.martin@gmail.com`

**Reply by Jacopodd** [reply]
Perfect, I sent you an email with WeTransfer

**Reply by martines3000** [reply]
I just watched the video and you did everything correct, so it should work, but it doesn't which is weird 😅.

Which browser are you using ? We are aware of some issues/problems on firefox.
We normally test using chromium browsers (chrome, brave, etc..).

**Reply by Jacopodd** [reply]
Yes, I use Google Chrome. What browser did you test on? So I text it on the same search engine

**Reply by martines3000** [reply]
If you are using a chromium based browser you can help us debug further.
Right click MetaMask extension and select `Manage Extension`, top right you can enable developer mode and then open the `background.html` link. This will open a developer console for MetaMask. Then you can repeat the flow and you can send a screenshot after you finish the last part and you get the error message. The console will probably include a better error message.
![image](https://github.com/blockchain-lab-um/masca/assets/35891136/4c363053-5ad1-428a-84a4-5b769092edd5)

**Reply by martines3000** [reply]
There is a possibility the request with Masca timeouts and thats why you get an error on the dapp, but I am not sure.

**Reply by martines3000** [reply]
You can ignore these errors: 
![image](https://github.com/blockchain-lab-um/masca/assets/35891136/8f253ea0-8cc2-462c-b749-6fbea2780d86)

**Reply by Jacopodd** [reply]
The first set of errors corresponds to the first time I click "Approve" on MetaMask, the second corresponds to the second "Approve". If you need anything else tell me.
![Screenshot (115)](https://github.com/blockchain-lab-um/masca/assets/80394163/c492b8dd-6a7b-4352-9794-cb7b4f7ac750)

**Reply by martines3000** [reply]
Are there any other errors that are not `premature close` or `at y...` ? Those are normal and not related to Masca.

**Reply by martines3000** [reply]
Can you also check the console on the Dapp (masca.io) ? Maybe we log a good error there, but I am not sure about that 😅.

**Reply by Jacopodd** [reply]
@martines3000 This is the error it gives me.
![Screenshot (116)](https://github.com/blockchain-lab-um/masca/assets/80394163/fc7ca8fc-6858-4ef0-ae2f-5f6fc85b97aa)

**Reply by martines3000** [reply]
> The first set of errors corresponds to the first time I click "Approve" on MetaMask, the second corresponds to the second "Approve". If you need anything else tell me.
> ![Screenshot (115)](https://github.com/blockchain-lab-um/masca/assets/80394163/c492b8dd-6a7b-4352-9794-cb7b4f7ac750)
> 
> 

1 more thing here. Can you check the network tab and see if any network request fail while handling the request? (This is inside the MetMask developer console).

**Reply by Jacopodd** [reply]
This is the result from the "Network" tab, there appear to be no errors.
![Screenshot (119)](https://github.com/blockchain-lab-um/masca/assets/80394163/784721cb-00f6-405f-b765-0028d6f1fb29)

**Reply by Jacopodd** [reply]
@martines3000 Can you tell me what the error is?

**Reply by martines3000** [reply]
Hi.

There are no errors in the network tab.

I'm still looking into it and what could be wrong. I didn't manage to replicate the issue. I tested it on multiple OS (windows, mac) and browsers (chrome, chromium and brave).

Now I am looking and debugging the library code to try see where it could go wrong and cause the error.

**Reply by Jacopodd** [reply]
Okay, I'm waiting for a reply. I'm pushing because I can make it work I can't finish the internship. Thank you for the attention you are giving me

**Reply by martines3000** [reply]
@Jacopodd 
Can you maybe try the flow on the [Brave browser](https://brave.com/) and get back to me with the results (if it works or if it still broken) ?

Also try to use the official MetaMask extension (not the Flask version) on `masca.io` as the latest Flask release has a bug that makes Masca unusable.

I know this is not a perfect solution, I am just interested to know if it is something specific to Chrome.
Are you maybe using a special version of chrome ?

**Reply by Jacopodd** [reply]
The Chrome version is: "Version 123.0.6312.106 (Build official) (a 64 bit)".
I tried removing Metamask flusk and installing normal Metamask, always trying on the internet but the result is the same.
I also tried with Brave and the result is identical. Below is a screenshot of the Brave console.
![Screenshot (121)](https://github.com/blockchain-lab-um/masca/assets/80394163/25f38842-d65b-4a20-9528-d846726ca9f0)

**Reply by martines3000** [reply]
Hey, I have 1 more idea what you can try. Maybe try the full flow while being connected to a VPN if you have one. Maybe there are some network communication errors happening.

**Reply by martines3000** [reply]
We also have a [Discord](https://discord.gg/9qhrWumf) that you can join and I can help you there. Maybe the communication will be easier/faster.

**Reply by martines3000** [reply]
Also, can you look again in the network tab of the MetaMask Developer console and screenshot what you get at this step ? 
![image](https://github.com/blockchain-lab-um/masca/assets/35891136/2240f2e6-8873-4c75-832c-024bb1b1ca7c)

The error happens right after this, so it is maybe connected.

**Reply by Jacopodd** [reply]
HI,
as for the VPN, I don't have it, maybe I know someone who could try it for me, but I need it for a few days.

As for the call on discord, I don't think it's feasible because I don't know how to speak English, I use Google Translate to exchange these messages.

In the meantime I'll send you what you asked about metamask
![Screenshot (122)](https://github.com/blockchain-lab-um/masca/assets/80394163/79b10a9b-5ccb-4d67-95f0-518ec4e84c75)

**Reply by Jacopodd** [reply]
The only thing different that we might have is that I don't have any verifiable credentials saved

**Reply by martines3000** [reply]
Thanks for this, it helps. I can see there is an error regarding timestamps. I will look into it and try to find a solution.

We have a "Masca Community Discord" where we also have a support/development text channel for providing help. It was just a suggestion, I don't want to force anything, so we can also keep the conversation here on GitHub.

**Reply by Jacopodd** [reply]
If we can always talk via message I would prefer, can you send me the link?

**Reply by martines3000** [reply]
Yes we can always chat over text. Here is the Discord invite https://discord.gg/9qhrWumf

**Reply by martines3000** [reply]
@Jacopodd 

I created a new release of Masca `1.2.3-beta.0` with some changes that will potentially solve your issue. You can test this new version using MetaMask Flask on `beta.masca.io`.

Also please check that you have the correct date, time and timezone set on your computer. I am mentioning this, because this was the only way I managed to replicate your issue, when my computer time and timezone were out of sync. This is important as Masca handles ID Token signing and a wrong `nbf` timestamp was added. 

Please respond with the results, if the new version works or not.

**Reply by Jacopodd** [reply]
Which channel do I write in?

In the meantime, to run version 1.2.3-beta.0 do I have to re-download the file from github or do I have to connect to masca.io as usual?

**Reply by martines3000** [reply]
Go to `beta.masca.io` and use `MetaMask Flask`.

You can write in `#Development`

**Reply by Jacopodd** [reply]
I don't see #Development
![Screenshot (128)](https://github.com/blockchain-lab-um/masca/assets/80394163/641d5ae0-51f4-418d-9a24-7d62a23b56c5)

**Reply by Jacopodd** [reply]
However I have tested beta.masca.io and everything seems to work. Next week I'll talk to the professor and let you know if I need anything else. Thank you very much for your availability

**Reply by martines3000** [reply]
Awesome that it works now 👍.

I added you the member role, now you should be able to see the channel.

I will also close this issue as resolved. If something breaks again, you can open a new one or ask us on Discord.
We are happy to help 😃.
